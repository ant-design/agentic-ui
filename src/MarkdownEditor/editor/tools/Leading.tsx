import { Anchor } from 'antd';
import { nanoid } from 'nanoid';
import React, { useCallback, useEffect, useRef } from 'react';

import { Node } from 'slate';
import { useDebounce, useGetSetState } from '../../index';
import { useEditorStore } from '../store';
import { getOffsetTop, slugify } from '../utils/dom';

type Leading = {
  title: string;
  level: number;
  id: string;
  key: string;
  dom?: HTMLElement;
  schema: object;
};

const cache = new Map<object, Leading>();

function buildTree(data: any[]) {
  let tree = {} as Record<string, any>;

  data.forEach((item: { level?: any }) => {
    if (item.level === 1) {
      tree = { ...item, children: [] };
    } else {
      let currentLevel = tree;
      for (let i = 2; i < item.level; i++) {
        if (!currentLevel?.children) {
          currentLevel.children = [];
        }
        if (currentLevel.children?.[currentLevel.children.length - 1]) {
          currentLevel =
            currentLevel.children?.[currentLevel.children.length - 1];
        }
      }
      if (!currentLevel?.children) {
        currentLevel.children = [];
      }
      currentLevel.children.push({ ...item, children: [] });
    }
  });

  return tree;
}

export const schemaToHeading = (schema: any) => {
  const headings: Leading[] = [];
  for (let s of schema) {
    if (s.type === 'head' && s.level <= 4) {
      const title = Node.string(s);
      const id = slugify(title);
      if (title) {
        headings.push({
          title,
          level: s.level,
          id,
          key: nanoid(),
          schema: s,
        });
      }
    }
  }

  return buildTree(headings)?.children?.map((h: any) => ({
    id: h.id,
    key: h.key,
    href: `#${h.id}`,
    children: h?.children?.map((subH: any) => ({
      id: subH.id,
      key: subH.key,
      href: `#${subH.id}`,
      title: subH.title,
    })),
    title: h.title,
    level: h.level,
  }));
};

/**
 * 配置次级标题的锚点
 */
export const TocHeading = ({ schema }: { schema: any }) => {
  const { store } = useEditorStore();
  const [state, setState] = useGetSetState({
    headings: [] as Leading[],
    active: '',
  });
  const box = useRef<HTMLElement>();
  const getHeading = useCallback(() => {
    if (schema?.length) {
      const headings: Leading[] = [];
      for (let s of schema) {
        if (s.type === 'head' && s.level <= 4) {
          if (cache.get(s)) {
            headings.push(cache.get(s)!);
            continue;
          }
          const title = Node.string(s);
          const id = slugify(title);
          if (title) {
            cache.set(s, {
              title,
              level: s.level,
              id,
              key: nanoid(),
              schema: s,
            });
            headings.push(cache.get(s)!);
          }
        }
      }
      setState({ headings });
    } else {
      setState({ headings: [] });
    }
  }, [schema]);

  useEffect(() => {
    cache.clear();
    getHeading();
    setState({ active: '' });
  }, [store?.container, store.editor.children]);

  useDebounce(getHeading, 100, [store.editor.children]);

  useEffect(() => {
    const div = box.current;
    if (div) {
      const scroll = (e: Event) => {
        const top = (e.target as HTMLElement).scrollTop;
        const container = store?.container;
        if (!container) return;
        const heads = state().headings.slice().reverse();
        for (let h of heads) {
          if (h.dom && top > getOffsetTop(h.dom, container) - 20) {
            setState({ active: h.id });
            return;
          }
        }
        setState({ active: '' });
      };
      div.addEventListener('scroll', scroll, { passive: true });
      return () => div.removeEventListener('scroll', scroll);
    }
    return () => {};
  }, []);

  if (!buildTree(state().headings).children) {
    return null as React.ReactNode;
  }
  return (
    <Anchor
      style={{
        minWidth: 200,
      }}
      offsetTop={64}
      items={buildTree(state().headings).children?.map((h: any) => ({
        id: h.id,
        key: h.key,
        href: `#${h.id}`,
        children: h?.children?.map((subH: any) => ({
          id: subH.id,
          key: subH.key,
          href: `#${subH.id}`,
          title: subH.title,
        })),
        title: h.title,
        level: h.level,
      }))}
    />
  );
};
