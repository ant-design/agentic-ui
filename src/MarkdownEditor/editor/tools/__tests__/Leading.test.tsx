import { render } from '@testing-library/react';
import { ConfigProvider } from 'antd';
import React from 'react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { Elements } from '../../../BaseMarkdownEditor';
import { TocHeading } from '../Leading';

// Mock the useEditorStore hook
vi.mock('../../store', () => ({
  useEditorStore: () => ({
    markdownEditorRef: { current: { children: [] } },
    markdownContainerRef: { current: null },
  }),
}));

// Mock useDebounce to prevent infinite loops
vi.mock('../../../BaseMarkdownEditor', async () => {
  const actual = await vi.importActual('../../../BaseMarkdownEditor');
  return {
    ...actual,
    useDebounce: (fn: any) => fn,
  };
});

describe('TocHeading Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  const mockSchema: Elements[] = [
    {
      type: 'head',
      level: 1,
      children: [{ text: 'Introduction' }],
    },
    {
      type: 'head',
      level: 2,
      children: [{ text: 'Getting Started' }],
    },
  ] as Elements[];

  it('renders table of contents correctly', () => {
    const { container, unmount } = render(
      <ConfigProvider
        theme={{
          hashed: false,
        }}
      >
        <TocHeading schema={mockSchema} />
      </ConfigProvider>,
    );
    const titles = container.querySelectorAll('.ant-anchor-link-title');
    expect(titles.length).toBe(2);
    expect(titles[0].textContent).toBe('Introduction');
    expect(titles[1].textContent).toBe('Getting Started');
    unmount();
  });

  it('renders nothing when no headers are present', () => {
    const { container, unmount } = render(
      <ConfigProvider
        theme={{
          hashed: false,
        }}
      >
        <TocHeading schema={[]} />
      </ConfigProvider>,
    );
    expect(container.firstChild).toBeNull();
    unmount();
  });

  it('handles nested headers correctly', () => {
    const nestedSchema: Elements[] = [
      {
        type: 'head',
        level: 1,
        children: [{ text: 'Chapter 1' }],
      },
      {
        type: 'head',
        level: 2,
        children: [{ text: 'Section 1.1' }],
      },
      {
        type: 'head',
        level: 2,
        children: [{ text: 'Section 1.2' }],
      },
      {
        type: 'head',
        level: 3,
        children: [{ text: 'Subsection 1.2.1' }],
      },
    ] as Elements[];

    const { container } = render(
      <ConfigProvider
        theme={{
          hashed: false,
        }}
      >
        <TocHeading schema={nestedSchema} />
      </ConfigProvider>,
    );
    const titles = container.querySelectorAll('.ant-anchor-link-title');
    expect(titles.length).toBe(4);
    expect(titles[0].textContent).toBe('Chapter 1');
    expect(titles[1].textContent).toBe('Section 1.1');
    expect(titles[2].textContent).toBe('Section 1.2');
    expect(titles[3].textContent).toBe('Subsection 1.2.1');
  });

  it('applies custom anchor props', () => {
    const customProps = {
      offsetTop: 100,
      style: { backgroundColor: 'red' },
    };

    const { container } = render(
      <ConfigProvider
        theme={{
          hashed: false,
        }}
      >
        <TocHeading schema={mockSchema} anchorProps={customProps} />
      </ConfigProvider>,
    );

    const anchor = container.querySelector('.ant-anchor');
    expect(anchor).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('applies max height and scrollbar styles', () => {
    const longSchema: Elements[] = Array(10)
      .fill(null)
      .map((_, index) => ({
        type: 'head',
        level: 1,
        children: [{ text: `Section ${index + 1}` }],
      })) as Elements[];

    const { container } = render(
      <ConfigProvider
        theme={{
          hashed: false,
        }}
      >
        <TocHeading schema={longSchema} />
      </ConfigProvider>,
    );
    const anchor = container.querySelector('.ant-anchor');
    expect(anchor).toBeDefined();
    expect(container).toMatchSnapshot();
  });
});
