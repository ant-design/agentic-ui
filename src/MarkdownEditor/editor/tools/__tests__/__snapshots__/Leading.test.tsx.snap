// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TocHeading Component > applies custom anchor props 1`] = `
<div>
  <div>
    <div
      class=""
    >
      <div
        class="ant-anchor-wrapper"
        style="max-height: calc(100vh - 100px); background-color: red;"
      >
        <div
          class="ant-anchor"
        >
          <span
            class="ant-anchor-ink"
          />
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#introduction"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Introduction
              </span>
            </a>
            <div
              class="ant-anchor-link"
            >
              <a
                class="ant-anchor-link-title"
                href="#getting-started"
                title=""
              >
                <span
                  style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
                >
                  Getting Started
                </span>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TocHeading Component > applies max height and scrollbar styles 1`] = `
<div>
  <div>
    <div
      class=""
    >
      <div
        class="ant-anchor-wrapper"
        style="max-height: min(calc(100vh - 180px), 300px); min-width: 200px; overflow-y: auto; overflow-x: hidden; padding-right: 4px;"
      >
        <div
          class="ant-anchor"
        >
          <span
            class="ant-anchor-ink"
          />
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-1"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 1
              </span>
            </a>
          </div>
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-2"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 2
              </span>
            </a>
          </div>
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-3"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 3
              </span>
            </a>
          </div>
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-4"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 4
              </span>
            </a>
          </div>
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-5"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 5
              </span>
            </a>
          </div>
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-6"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 6
              </span>
            </a>
          </div>
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-7"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 7
              </span>
            </a>
          </div>
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-8"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 8
              </span>
            </a>
          </div>
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-9"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 9
              </span>
            </a>
          </div>
          <div
            class="ant-anchor-link"
          >
            <a
              class="ant-anchor-link-title"
              href="#section-10"
              title=""
            >
              <span
                style="max-width: 200px; display: inline-block; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; font-size: 14px;"
              >
                Section 10
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
